### PR TITLE
Fixes compiler feature detection and config flags using CMake

### LINKED ISSUE #837: HAVE_GC might remain undefined in 2.6.3 (if built with cmake)
In Flint 2.5* (and even later, maybe, but not in 2.6.3) the macro `HAVE_GC` was always defined, and set to 0 or 1. Now
if Flint is built with `cmake`, then instead of 
```
$ grep -R HAVE_GC
flint-config.h:#define HAVE_GC 0
```
one just sees no HAVE_GC in  `flint-config.h`. This is a kind of a bug.

----------------

It is in fact just not possible to tell cmake to use GC with Flint, as of Flint 2.6.3 or earlier.

### PR DESCRIPTION
**Pull request by mahrud**
Fixes #837

### CODE REVIEW

**Comment by wbhart** [comment]
Thank you. I'll merge this tomorrow and make sure it is backported to the 2.7 release branch.

**Comment by mahrud** [comment]
@isuruf let me know if you have any suggestions on this.

**Review by isuruf — COMMENTED** [review]
Looks good to me. Do we want Apache-2.0 code (`FindCBLAS.cmake`)?

**Comment by mahrud** [comment]
> Looks good to me. Do we want Apache-2.0 code (`FindCBLAS.cmake`)?

Good point, I rewrote it from scratch.

File: CMake/FindCBLAS.cmake [hunk 1]
```diff
@@ -0,0 +1,28 @@
...
+#  CBLAS_INCLUDE_DIRS - the header directory containing cblas.h
+#  CBLAS_LIBRARIES    - the CBLAS library
+
+# Copyright (c) 2020, Mahrud Sayrafi, <mahrud@umn.edu>
+# Redistribution and use is allowed according to the terms of the BSD license.
+
+find_path(CBLAS_INCLUDE_DIRS NAMES cblas.h
+  HINTS CBLAS_ROOT ENV CBLAS_ROOT
+  PATHS ${INCLUDE_INSTALL_DIR} ${CMAKE_INSTALL_PREFIX}/include
+  PATH_SUFFIXES openblas cblas blis
```
Comment (isuruf) [inline CMake/FindCBLAS.cmake:14]:
```suggestion
```

Comment (mahrud) [reply CMake/FindCBLAS.cmake:14]: I'm afraid this is necessary, for instance on my machine:
```
[mahrud@noether emacs]$ ls /usr/include/*/cblas.h 
/usr/include/cblas/cblas.h  /usr/include/openblas/cblas.h
```

**Comment by mahrud** [comment]
Something seems to be wrong with detecting popcnt on windows. I unfortunately can't test on windows, but if anybody has suggestions for resolving it, feel free to push commits directly on this branch.

**Comment by wbhart** [comment]
Would @BrianGladman be able to help with detecting popcnt on Windows, or is this a CMake specific issue?

**Comment by BrianGladman** [comment]
I can write a bit of C code to detect whether the POPCNT instruction is present on the machine running the code if that is what you need.

**Comment by mahrud** [comment]
Actually, I think the error was due to a typo in my code. Should be fixed now!

File: CMake/FindCBLAS.cmake (hunk 1 above)
Comment (mahrud) [reply CMake/FindCBLAS.cmake:14]: Also, I believe CMake checks to see if the file is on the path alone, and if not tries the suffixes in order.

**Comment by mahrud** [comment]
The windows build compiled, but a large number of tests failed with a segmentation fault. Are there any windows tests using the configure system for comparison?

@BrianGladman would you be able to look into what is happening? It would help if you could upload the `flint-config.h` or `config.log` files produced by both `configure` and `cmake` on windows.

**Comment by wbhart** [comment]
I know @BrianGladman got everything to work with MSVC on his system. So hopefully he'll be able to help.

I'm not sure if this log contains an example of building with configure on Windows:

https://ci.appveyor.com/project/wbhart/flint2/builds/36743521/job/x8k100ii36c72acv

I think it used to do both a configure build and a CMake build. But I'm not sure if that is so any more.

**Comment by BrianGladman** [comment]
@mahrud  This also happened to me when I tried to build my Visual Studio 2019 build a couple of days ago. It turned out that some defines had been changed by adding FLINT_ in front of them.   The ones I had to modify in my own config.h file are:

-----------------------------------------------------------------------------------------
/* Define if the compiler supports and should use thread-local storage */
#define FLINT_USES_TLS 1

/* Define if the library should be thread-safe, no matter whether FLINT_USES_TLS is used */
#define FLINT_REENTRANT 1

/* ./configure --enable-assert option, to enable some ASSERT()s */
#undef FLINT_WANT_ASSERT

/* Define if you cpu_set_t in sched.h */
#define FLINT_USES_CPUSET 1

#define FLINT_USES_PTHREAD 1
-----------------------------------------------------------------------------------------

I can't supply a cmake config.h file or a configure based one as my build uses its own config,h file.  But you can find the one I use in my repository at https://github.com/BrianGladman/flint where it is in the build.vc sub-directory.   This is copied to flint-config.h in my prebuild step.  I hope this helps, if not do get back to me.

**Comment by wbhart** [comment]
The latest build indicates the issue is just MSVC specific now.

**Comment by wbhart** [comment]
Ignore me. I see that's been the case for a while now.

**Comment by mahrud** [comment]
Yes, only using cl on windows seems to have this issue. Though in the last commit I tried to manually disable fenv and that broke other things.

I'm not sure what's the problem, to be honest, but it probably has to do with either TLS or CPUSET not being checked correctly with cl.

**Comment by wbhart** [comment]
I don't know anything at all about CMake, but this looks very much like it is gcc specific:

```
check_c_source_runs([[int main(int argc, char ** argv) {
  #if defined(_WIN64)
  return __builtin_popcountll(argc) == 100;
  #else
  return __builtin_popcountl(argc) == 100;
  #endif
  }]] FLINT_USES_POPCNT)
unset(CMAKE_REQUIRED_FLAGS)
```

Is that ok?

**Comment by mahrud** [comment]
That's the same code that configure uses to detect builtin popcount. Comparing a previously successful build (like [this one](https://ci.appveyor.com/project/wbhart/flint2/builds/36761902/job/675nsoqk12tk52jc?fullLog=true)) and the current build for COMPILER=cl, ARCH=64, here are the differences:

Before:
```
-- Performing Test HAS_FLAG_MPOPCNT
-- Performing Test HAS_FLAG_MPOPCNT - Failed
-- Performing Test HAS_FLAG_UNROLL_LOOPS
-- Performing Test HAS_FLAG_UNROLL_LOOPS - Failed
-- Looking for pthread_getaffinity_np in C:/Miniconda36-x64/Library/lib/pthreads.lib
-- Looking for pthread_getaffinity_np in C:/Miniconda36-x64/Library/lib/pthreads.lib - not found
```
Then, `FLINT_USES_CPUSET` was being set depending on `pthread_getaffinity_np`, and `FLINT_USES_TLS` was on by default.

Now:
```
-- Performing Test HAS_FLAG_MPOPCNT
-- Performing Test HAS_FLAG_MPOPCNT - Failed
-- Performing Test HAS_FLAG_UNROLL_LOOPS
-- Performing Test HAS_FLAG_UNROLL_LOOPS - Failed
-- Performing Test FLINT_USES_POPCNT
-- Performing Test FLINT_USES_POPCNT - Failed
-- Performing Test FLINT_USES_FENV
-- Performing Test FLINT_USES_FENV - Success
-- Performing Test FLINT_USES_CPUSET
-- Performing Test FLINT_USES_CPUSET - Failed
-- Performing Test FLINT_USES_TLS
-- Performing Test FLINT_USES_TLS - Failed
```
Previously a bunch of things were just not being tested or not set, but almost all of them are not being detected now anyway, so I think most of the behavior is the same. The only one now being set to 1 is `FLINT_USES_FENV`, but I tried disabling that manually and it didn't help.

On the other hand, `FLINT_USES_TLS` is now being set to 0, which I didn't realize until now that it was previously set to 1 by default.

Is this code also gcc specific?
```c
__thread int x = 42; int main(int argc, char ** argv) { return x != 42; }
```

**Comment by mahrud** [comment]
The tests are passing! And also now I see this code in `flint.h`:
```c
#if FLINT_USES_TLS
#if __STDC_VERSION__ >= 201112L
#define FLINT_TLS_PREFIX _Thread_local
#elif defined(_MSC_VER)
#define FLINT_TLS_PREFIX __declspec(thread)
#elif defined(__GNUC__)
#define FLINT_TLS_PREFIX __thread
#else
#error "thread local prefix defined in C11 or later"
#endif
#else
#define FLINT_TLS_PREFIX
#endif
```
I think I can properly check this using:
```cmake
# Thread-local storage configuration
if(cxx_thread_local IN_LIST CMAKE_CXX_COMPILE_FEATURES)
  set(FLINT_USES_TLS ON CACHE BOOL "Use thread local storage.")
endif()
```

**Comment by mahrud** [comment]
To add to the confusion, now one check on linux using configure is failing:
```
    CC   build/test/t-sdiv_qrnnd
/home/travis/build/wbhart/flint2/libflint.so: undefined reference to `__gmpn_neg'
/home/travis/build/wbhart/flint2/libflint.so: undefined reference to `__gmpn_com'
collect2: error: ld returned 1 exit status
```
Here is the difference between when it passed:
```
checking if gmp.h version and libgmp version are the same... (6.1.0/6.1.0) yes
checking for __gmpn_sbpi1_divappr_q... yes
checking for __gmpn_invert_limb... yes
checking for __gmpn_rsblsh_n... yes
```
and when now:
```
checking if gmp.h version and libgmp version are the same... (6.0.0/6.0.0) yes
checking for __gmpn_sbpi1_divappr_q... no
checking for __gmpn_invert_limb... no
checking for __gmpn_rsblsh_n... no
```
I don't think this is related to my changes, but might be worth opening a new issue.

**Comment by mahrud** [comment]
Other than the configure errors, one test timed out on COMPILER=cl, ARCH=64:
```
	1200 - fmpz_mod_mat-test-t-mul_classical_threaded (Timeout)
```

**Comment by wbhart** [comment]
The linux failure is unrelated to this PR. It can be ignored for now. It randomly passes on that machine. No idea why.

**Comment by wbhart** [comment]
The timeout is weird. We can try it again.

**Comment by mahrud** [comment]
> The timeout is weird. We can try it again.

The only other change that may have caused this is the few `#if HAVE_PTHREAD`s that I changed in mpolyu.c, I can revert that to see if it helps, but it might be worth you looking into whether that's a bug.

**Comment by mahrud** [comment]
Great, it succeeded! (with the exception of the gmp failure)
This PR is ready I guess, but I suspect there's a bug in mpolyu.c!

**Comment by wbhart** [comment]
Yes, you are right, there is definitely a bug in mpolyu.c. @tthsqe12 is going to take a look at that on Monday.

Thanks for getting the rest working. That's great!

**Comment by tthsqe12** [comment]
Yes, those HAVE_PTHREAD should all be FLINT_USES_PTHREAD, or whatever it got renamed to. This is my bad, as I must have I messed it up when I was fixing merge conflicts.

**Comment by wbhart** [comment]
@tthsqe12 Right, but then the test hangs!

**Comment by tthsqe12** [comment]
Then, it is probably a ptoblem with the order of includes. Didn't we have a similar problem last time?

**Comment by wbhart** [comment]
Surely that would only affect the compilation, not the running of the test?

**Comment by mahrud** [comment]
> Its gcc on linux that is failing?

Is actually only 32 bit windows with MSVC that times out on a classical_threaded test after 10 minutes.

**Comment by wbhart** [comment]
@tthsqe12 The failure you can see right now is some random GMP linking issue. I think it can be ignored.

The issue we are talking about happens when those defines are set correctly. They aren't set correctly right now, which is why you don't see a failure.

**Review by tthsqe12 — COMMENTED** [review]
@mahrud I tried this on appveyor with FLINT_USES_PTHREAD in all locations and the only conclusion that I come to is that the fft tests are randomly interrupting fmpz_mod_mat/t-mul_classical_threaded on that cl-32 setup. fmpz_mod_mat/t-mul_classical was not being tested on this cl-32 setup before, and test takes a while to complete even on a fully capable machine. The cl-32 machine is already _very slow_ and this along with the threaded fft
test is just too much sometimes. Lets hope that @wbhart finds a way to tone
down the fft tests. In the mean time this should also help in fmpz_mod_mat/t-mul_classical_threaded.c, which is standard practice elsewhere in the test code:
```
int
main(void)
{
#if FLINT_USES_PTHREAD && (FLINT_USES_TLS || FLINT_REENTRANT)
    slong i, max_threads = 5;
    slong tmul = 1000;
#ifdef _WIN32
    tmul = 100;
#endif
#endif
    FLINT_TEST_INIT(state);

    flint_printf("mul_classical_threaded....");
    fflush(stdout);

#if FLINT_USES_PTHREAD && (FLINT_USES_TLS || FLINT_REENTRANT)
    for (i = 0; i < tmul * flint_test_multiplier(); i++)
    {
        fmpz_mod_mat_t A, B, C, D;
```

**Comment by wbhart** [comment]
There isn't really much I can do with the FFT tests. I've already cut them back as far as I'm comfortable with.

Perhaps we should just increase the timeout. I'll be glad when 32 bits finally dies.

**Comment by mahrud** [comment]
Even 150 seconds without fft tests seems too long! The t-mul_classical_threaded test took a fraction of a second on the [last build from trunk](https://ci.appveyor.com/project/wbhart/flint2/builds/36761902/job/d6rkeosuox0fvx3f):
```
1199/2028 Test #1200: fmpz_mod_mat-test-t-mul_classical_threaded ............................   Passed    0.04 sec
```

Let me know what you'd like me to do in this PR. I can revert the last commit, or leave it as is for now.

**Comment by tthsqe12** [comment]
Revert the last commit and make my recommended changes to fmpz_mod_mat/t-mul_classical_threaded.c. Maybe start at tmul=20, and then increase it untill the test takes about 10 seconds on that delinquent cl32 box.

**Comment by mahrud** [comment]
Since it's not CMake related (and the current state of this PR doesn't
change anything not CMake related), could you do that in a separate PR? I
probably won't have time to work on it this week.

**Comment by tthsqe12** [comment]
Ok

**Comment by wbhart** [comment]
Ok, I'll merge this PR as is. Thanks for the help @mahrud !

**Pull request merged**